### ISSUE TITLE
Pub/Sub topic and subscription creation timeout

### ISSUE THREAD

**Issue by lacarvalho91** [TheKevJames/tools#343]
The issue I am seeing is that `pubsubc` fails to create topics + subscriptions when the container starts up, and when this happens I can just see `Operation timed out` in the logs. When this happens it doesn't retry or anything. I'm guessing this happens when the server takes too long to start. [Here](https://github.com/TheKevJames/tools/blob/1b86605674942cf107a39e3122291674b63dd0be/docker-gcloud-pubsub-emulator/run.sh#L8) the pubsub client is run in the background before the actual server starts so I guess there is a race.

If I run `pubsubc` manually after the topics are created fine.

I'm running the image in `minikube` using [onechart](https://github.com/gimlet-io/onechart/tree/v0.34.0).

**Comment by TheKevJames** [comment]
Can't track this one down at all! As far as I can tell, the issues started with `gcloud v407`: any version greater than that is causing issues.

I've updated the autobuild script to attach the gcloud version number, so pinning to `thekevjames/gcloud-pubsub-emulator:406.0.0` will let folks avoid this issue, but no dice on figuring out a real fix.

If anyone can make use of a more recent tag than `406.0.0` and figure out what's going on here, or even provide more info to help with debugging, that'd be much appreciated!

**Comment by lacarvalho91** [comment]
@TheKevJames sorry I wasn't checking updates to this. I've tried out 420.0.0 but the problem I'm seeing now is that the readiness port that gets opened up when the topics and subscriptions have been created is not responding to requests. Any requests on that port just hang indefinitely, this even happens when I exec into the container and `curl localhost:8682`.

In the logs I see:
```
configuring project: my-proj
- creating topic: my-topic
  - creating subscription: my-sub
Done building projects/topics/subscriptions! Opening readiness port...
```

So the topic and sub are getting created, but then when I do the curls no response comes back and I can see in the logs:
```
GET / HTTP/1.1
Host: 172.17.0.5:8682
User-Agent: kube-probe/1.23
Accept: */*
Connection: close

GET / HTTP/1.1
Host: localhost:8682
User-Agent: curl/7.74.0
Accept: */*
```

So I can see it is logging k8s readiness probe requests and the requests I am sending from within the container.

**Comment by lacarvalho91** [comment]
I've also just seen the originally problem happen again, only seems to happen when we run the container on our Jenkins instance though:
```
Executing: /usr/lib/google-cloud-sdk/platform/pubsub-emulator/bin/cloud-pubsub-emulator --host=0.0.0.0 --port=8681
[pubsub] This is the Google Pub/Sub fake.
[pubsub] Implementation may be incomplete or differ from the real system.
[pubsub] Apr 12, 2023 10:46:35 AM com.google.cloud.pubsub.testing.v1.Main main
[pubsub] INFO: IAM integration is disabled. IAM policy methods and ACL checks are not supported
[pubsub] SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
[pubsub] SLF4J: Defaulting to no-operation (NOP) logger implementation
[pubsub] SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
Operation timed out
```

I think it would be a good idea to add some resiliency to calls that are creating the topics and subscriptions. As I'm pretty sure that in this case a simple retry would do the trick.

**Comment by TheKevJames** [comment]
Interesting! I've so far been unable to replicate the issues you're reporting above, so any additional information you can find would be very helpful!

**Comment by TheKevJames** [comment]
Closing for lack of response, please feel free to reopen if you're still experiencing this, especially if you have any more debugging info to add!

**Comment by filipeherculano** [comment]
Still happens for me. I'm on latest version and I'm using it on docker compose

```
  google-pubsub-emulator:
    image: 'thekevjames/gcloud-pubsub-emulator'
    environment:
      PUBSUB_PROJECT1: "project,topic:sub"
    ports:
      - '8681:8681'
    networks:
      - local_network
```

**Comment by shintasmith** [comment]
Hello, @TheKevJames 
We are seeing this issue with the latest image. It was working about ~2 weeks ago. Or maybe we just got lucky then. 

We went back and tried version `497.0.0` and `496.0.0` -- they all have the same issue: the project/topic/subscription are not created. We saw a message:
```
Operation timed out
```
The issue is intermittent. Some times it works. But often times it does not. 

Looking at the documentation for `wait-for`, it looks like the default timeout is 15s. Maybe 15s is too short? We will be building the image locally and testing it with 60s timeout on `wait-for`. 

If that works, then I can submit a PR to allow the `run.sh` to check for an environment variable to control the timeout. If the environment variable does not exist, it will use whatever is the default timeout for `wait-for`. If the environment variable exists, then it will use the value for the timeout. I can open a PR shortly.
